iFixit/react-commerce
# Playwright: Abstract Handler Creation for MSW

## Description

In #1275 we created the ability to set dynamic handlers for `MSW`. However, the API for creating a handler is rather bloated and a lot of it can be extracted to reduce the boilerplate.

As such, we created a new static helper function to create these handlers.

## CR:

Commit by commit would be the easiest way to understand the code as some obscure context is needed in some cases.

Overall, the end goal was to abstract creating these dynamic handlers just enough to reduce the boilerplate and still be simple to use. Most of this was accomplished by making use of Typescript's features. 

## QA:

Besides checking CI still passing, I am not sure if it will be worth validating if the error handling works, as that might be exhaustive. 

I already confirmed that the `international buybox` handler still works using similar steps as https://github.com/iFixit/react-commerce/pull/1319#issuecomment-1416451999.

I also confirmed that the type checking is working as expected for the overloaded signatures:


|Before|After|
|:---:|:---:|
|  ![GraphQL body type is allowed](https://user-images.githubusercontent.com/22064420/217672866-e36d2476-6155-44df-ac17-bea9889c653b.png) | ![Graphql only accepts an object not a string](https://user-images.githubusercontent.com/22064420/217672913-be2a3a05-58be-4f01-87e1-c6abc59485db.png)|
<p align=center> <strong>In the scenario above, a GraphQLHandler requires a body that is only of type <code>Object</code></strong></p>


cr_req 2
qa_req 0

Connects: #1290
